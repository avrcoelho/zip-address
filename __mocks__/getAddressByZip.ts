const getAddressByZip = {
	cep: "02050-010",
	logradouro: "Rua Miguel Mentem",
	complemento: "",
	bairro: "Vila Guilherme",
	localidade: "SÃ£o Paulo",
	uf: "SP",
	unidade: "",
	ibge: "3550308",
	gia: "1004"
};

export default getAddressByZip;