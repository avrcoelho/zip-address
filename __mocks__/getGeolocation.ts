const getGeolocation = {
	"results": [{
		"address_components": [{
				"long_name": "02050",
				"short_name": "02050",
				"types": [
					"postal_code"
				]
			},
			{
				"long_name": "Marshfield",
				"short_name": "Marshfield",
				"types": [
					"locality",
					"political"
				]
			},
			{
				"long_name": "Plymouth County",
				"short_name": "Plymouth County",
				"types": [
					"administrative_area_level_2",
					"political"
				]
			},
			{
				"long_name": "Massachusetts",
				"short_name": "MA",
				"types": [
					"administrative_area_level_1",
					"political"
				]
			},
			{
				"long_name": "United States",
				"short_name": "US",
				"types": [
					"country",
					"political"
				]
			}
		],
		"formatted_address": "Marshfield, MA 02050, USA",
		"geometry": {
			"bounds": {
				"northeast": {
					"lat": 42.2037299,
					"lng": -70.63502799999999
				},
				"southwest": {
					"lat": 42.0525898,
					"lng": -70.7879798
				}
			},
			"location": {
				"lat": 42.110535,
				"lng": -70.70495199999999
			},
			"location_type": "APPROXIMATE",
			"viewport": {
				"northeast": {
					"lat": 42.2037299,
					"lng": -70.63502799999999
				},
				"southwest": {
					"lat": 42.0525898,
					"lng": -70.7879798
				}
			}
		},
		"partial_match": true,
		"place_id": "ChIJnZhyCi6n5IkRmQ_yKsX6weY",
		"types": [
			"postal_code"
		]
	}],
	"status": "OK"
};

export default getGeolocation;