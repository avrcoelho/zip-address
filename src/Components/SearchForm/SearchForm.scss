@import '../../Scss/root';

.search-form {
	border-radius: $border-radius;
	margin-bottom: 2em;
	padding: 2em;

	&__container {
		@media screen and (max-width: $sm-beakpoint) { flex-direction: column }
		align-items: center;
		border: none;
		display: flex;
	}

	&__title {
		@media screen and (max-width: $sm-beakpoint) { text-align: center }
		margin-top: 0;
	}

	&__input {
		@media screen and (max-width: $sm-beakpoint) { margin: .7em 0 }
		border-radius: $border-radius;
		border: 1px solid $primary-color;
		flex: 1;
		margin: 0 .7em;
		outline: none;
		padding: .5em;

		&:focus { border-color: lighten($primary-color, 20%) }

		&-container {
			align-self: stretch;
			display: flex;
			flex-direction: column;
			justify-content: stretch;
			position: relative;
		}

		&-message {
			@media screen and (min-width: $sm-beakpoint + 1) {
				bottom: -2.3em;
				left: 1em;
				position: absolute;
			}
			color: red;
			margin-bottom: .7em;

			&:empty { display: none }
		}

		&--error { border-color: red; }
	}
}