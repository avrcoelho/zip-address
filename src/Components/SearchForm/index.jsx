import React, { useState } from 'react';
import MaskedInput from 'react-text-mask';
import { Formik } from 'formik';
import { CSSTransition } from 'react-transition-group';

import { useAppContext } from '../../Providers/App.Context';
import { getAddressByZip, getGeolocation } from '../../Utils/API';

import AddressInfo from '../AddressInfo';
import Button from '../Button';
import { Search } from '../Icons';

import './SearchForm.scss';

export default function SearchForm() {
	const { loading, setLoading } = useAppContext();
	const [ map, setMap ] = useState({
		visible: false,
		coords: {
			lat: 0,
			lng: 0
		},
		address: {
			logradouro: '',
			bairro: '',
			localidade: '',
			uf: '',
			cep: ''
		}
	});
	const [ hasError, setHasError ] = useState(false);

	return (

		<>
			{/* FORM WITH VALIDATION */}
			<Formik
				initialValues={{zipCode: ''}}
				validate={values => {
					let errors = {}

					if(!/\d{5}-\d{3}/.test(values.zipCode)) errors.zipCode = 'Digite um CEP válido';

					return errors;
				}}
				onSubmit={async ({ zipCode }, { resetForm }) => {
					if(map.address.cep === zipCode) return resetForm();

					setMap(prevState => ({
						...prevState,
						visible: false
					}));
					setHasError(false);

					try {
						setLoading(true);

						const address = await getAddressByZip(zipCode);
						const Geolocation = await getGeolocation(address.logradouro);
						const latLong = Geolocation.results[0].geometry.location;

						setMap({
							visible: true,
							coords: {
								...latLong
							},
							address
						});

						setLoading(false);
						resetForm();
					} catch(err) {
						setHasError(true);
						setLoading(false);
					}
				}}
			>
				{({
					values,
					errors,
					touched,
					handleChange,
					handleSubmit
				}) => (
					<form className="search-form bg-secondary color-primary" onSubmit={handleSubmit}>
						<h2 className="search-form__title">Consultar endereço</h2>

						<div className="search-form__container">
							<label htmlFor="zipCode">
								CEP:
							</label>

							<span className="search-form__input-container">
								<MaskedInput
									autoComplete="off"
									className={`${(errors.zipCode && touched.zipCode) ? 'search-form__input--error ' : ''}search-form__input`}
									guide={true}
									id="zipCode"
									mask={[/\d/, /\d/, /\d/, /\d/, /\d/, '-', /\d/, /\d/, /\d/]}
									name="zipCode"
									onChange={handleChange}
									placeholder="_____-___"
									type="tel"
									value={values.zipCode}
								/>
								{errors.zipCode && <small className="search-form__input-message search-form__input-message--error">{errors.zipCode && touched.zipCode && errors.zipCode}</small>}
							</span>

							<Button
								big
								className="search-form__button"
								type="submit"
								disabled={loading}
							>
								<Search />
							</Button>
						</div>
					</form>
				)}
			</Formik>

			{/* INFO AND MAP */}
			<CSSTransition
				in={map.visible}
				timeout={300}
				classNames="address-info"
				unmountOnExit
			>
				<AddressInfo
					street={map.address.logradouro}
					neighborhood={map.address.bairro}
					city={map.address.localidade}
					uf={map.address.uf}
					zip={map.address.cep}
					geolocation={map.coords}
				/>
			</CSSTransition>

			{/* ERROR HANDLING */}
			{hasError && <p className="text-center"><strong>Ocorreu um erro. <br /> Verifique se o CEP digitado é válido, ou, tente novamente mais tarde.</strong></p>}
		</>
	)
}
