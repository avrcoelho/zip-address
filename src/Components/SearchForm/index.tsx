import  * as React from 'react';
import MaskedInput from 'react-text-mask';
import { Formik } from 'formik';
import { CSSTransition } from 'react-transition-group';

import { useAppContext } from '../../Providers/App.Context';
import { getAddressByZip, getGeolocation } from '../../Utils/API';

import AddressInfo from '../AddressInfo';
import Button, { ButtonTypes } from '../Button';
import { Search } from '../Icons';

import './SearchForm.scss';

export default function SearchForm() {
	// Loading from context
	const { loading, setLoading } = useAppContext();

	// Map State (Infos and visibility hanlder)
	const [ map, setMap ] = React.useState({
		visible: false,
		coords: {
			lat: 0,
			lng: 0
		},
		address: {
			logradouro: '',
			bairro: '',
			localidade: '',
			uf: '',
			cep: ''
		}
	});

	// Error state visibility handler
	const [ hasError, setHasError ] = React.useState(false);

	return (
		<>
			{/* FORM WITH VALIDATION */}
			<Formik
				initialValues={{zipCode: ''}}
				onSubmit={async ({ zipCode }, { resetForm }) => {
					if(map.address.cep === zipCode) return resetForm();

					setMap(prevState => ({
						...prevState,
						visible: false
					}));
					setHasError(false);

					try {
						setLoading(true);

						const address = await getAddressByZip(zipCode);
						const Geolocation = await getGeolocation(address.logradouro);
						const latLong = Geolocation.results[0].geometry.location;

						setMap({
							visible: true,
							coords: {
								...latLong
							},
							address
						});

						setLoading(false);
						resetForm();
					} catch(err) {
						setHasError(true);
						setLoading(false);
					}
				}}
				validate={values => {
					let errors = { zipCode: '' }

					if(!/\d{5}-\d{3}/.test(values.zipCode)) errors.zipCode = 'Digite um CEP válido';

					return errors;
				}}
				validateOnBlur
			>
				{({
					values,
					errors,
					touched,
					handleChange,
					handleSubmit
				}) => (
					<form className="search-form bg-secondary color-primary" onSubmit={handleSubmit}>
						<h2 className="search-form__title">Consultar endereço</h2>

						<div className="search-form__container">
							<label htmlFor="zipCode">
								CEP:
							</label>

							<span className="search-form__input-container">
								<MaskedInput
									autoComplete="off"
									className={`${(errors.zipCode && touched.zipCode) ? 'search-form__input--error ' : ''}search-form__input`}
									guide={true}
									id="zipCode"
									mask={[/\d/, /\d/, /\d/, /\d/, /\d/, '-', /\d/, /\d/, /\d/]}
									name="zipCode"
									onChange={handleChange}
									placeholder="_____-___"
									type="tel"
									value={values.zipCode}
								/>
								{errors.zipCode && <small className="search-form__input-message search-form__input-message--error">{errors.zipCode && touched.zipCode && errors.zipCode}</small>}
							</span>

							<Button
								big
								className="search-form__button"
								type={ButtonTypes.submit}
								disabled={loading}
							>
								<Search />
							</Button>
						</div>
					</form>
				)}
			</Formik>

			{/* INFO AND MAP */}
			<CSSTransition
				in={map.visible}
				timeout={300}
				classNames="address-info"
				unmountOnExit
			>
				<AddressInfo
					street={map.address.logradouro}
					neighborhood={map.address.bairro}
					city={map.address.localidade}
					uf={map.address.uf}
					zip={map.address.cep}
					geolocation={map.coords}
				/>
			</CSSTransition>

			{/* ERROR HANDLING */}
			{hasError && <p data-test="error-message" className="text-center"><strong>Ocorreu um erro. <br /> Verifique se o CEP digitado é válido, ou, tente novamente mais tarde.</strong></p>}
		</>
	)
}