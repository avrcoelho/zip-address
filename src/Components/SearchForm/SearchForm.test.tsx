import * as React from 'react';
import * as wait from 'waait';
import { mount } from 'enzyme';

import getAddressByZip from '../../../__mocks__/getAddressByZip';
import getGeolocation from '../../../__mocks__/getGeolocation';

import AppProvider from '../../Providers/App.Context'

import SearchForm from './index';

describe('<SearchForm />', () => {
	it('should mount properly', () => {
		const wrapper = mount(
			<AppProvider>
				<SearchForm />
			</AppProvider>
		);

		expect(wrapper).toMatchSnapshot();
	});

	it('should fill input', () => {
		const wrapper = mount(
			<AppProvider>
				<SearchForm />
			</AppProvider>
		);

		const inputValue = '00000-000'

		wrapper
			.find('input#zipCode')
			.simulate('change', {
				persist: () => {},
				target: { value: inputValue, name: 'zipCode' }
			});

		const newValue = wrapper.find('input#zipCode').props().defaultValue

		expect(newValue).toBe(inputValue);
	});

	it('should submit form', async () => {
		fetchMock
			.once(JSON.stringify(getAddressByZip))
			.once(JSON.stringify(getGeolocation))

		const wrapper = mount(
			<AppProvider>
				<SearchForm />
			</AppProvider>
		);

		const inputValue = '00000-000'

		wrapper
		.find('input#zipCode')
		.simulate('change', {
			persist: () => {},
			target: { value: inputValue, name: 'zipCode' }
		});

		wrapper
			.find('form')
			.simulate('submit');

		await wait();
		wrapper.update();

		expect(wrapper.find('AddressInfo')).toHaveLength(1);
	});

	it('should submit form and get errors properly', async () => {
		fetchMock
			.mockRejectOnce()

		const wrapper = mount(
			<AppProvider>
				<SearchForm />
			</AppProvider>
		);

		const inputValue = '00000-000'

		wrapper
		.find('input#zipCode')
		.simulate('change', {
			persist: () => {},
			target: { value: inputValue, name: 'zipCode' }
		});

		wrapper
			.find('form')
			.simulate('submit');

		await wait();
		wrapper.update();

		expect(wrapper.find('[data-test="error-message"]')).toHaveLength(1);
	});
});