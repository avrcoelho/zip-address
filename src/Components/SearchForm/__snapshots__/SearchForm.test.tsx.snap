// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SearchForm /> should mount properly 1`] = `
<AppProvider>
  <SearchForm>
    <Formik
      enableReinitialize={false}
      initialValues={
        Object {
          "zipCode": "",
        }
      }
      isInitialValid={false}
      onSubmit={[Function]}
      validate={[Function]}
      validateOnBlur={true}
      validateOnChange={true}
    >
      <form
        className="search-form bg-secondary color-primary"
        onSubmit={[Function]}
      >
        <h2
          className="search-form__title"
        >
          Consultar endere√ßo
        </h2>
        <div
          className="search-form__container"
        >
          <label
            htmlFor="zipCode"
          >
            CEP:
          </label>
          <span
            className="search-form__input-container"
          >
            <t
              autoComplete="off"
              className="search-form__input"
              guide={true}
              id="zipCode"
              mask={
                Array [
                  /\\\\d/,
                  /\\\\d/,
                  /\\\\d/,
                  /\\\\d/,
                  /\\\\d/,
                  "-",
                  /\\\\d/,
                  /\\\\d/,
                  /\\\\d/,
                ]
              }
              name="zipCode"
              onChange={[Function]}
              placeholder="_____-___"
              render={[Function]}
              type="tel"
              value=""
            >
              <input
                autoComplete="off"
                className="search-form__input"
                defaultValue=""
                id="zipCode"
                name="zipCode"
                onBlur={[Function]}
                onChange={[Function]}
                placeholder="_____-___"
                type="tel"
              />
            </t>
          </span>
          <Button
            big={true}
            className="search-form__button"
            disabled={false}
            type="submit"
          >
            <button
              className="button--big search-form__button button bg-primary color-secondary"
              disabled={false}
              type="submit"
            >
              <span
                className="button__content"
              >
                <Search>
                  <span
                    aria-label="Busca"
                    role="img"
                  >
                    <MdSearch>
                      <IconBase
                        attr={
                          Object {
                            "viewBox": "0 0 24 24",
                          }
                        }
                      >
                        <svg
                          fill="currentColor"
                          height="1em"
                          stroke="currentColor"
                          strokeWidth="0"
                          style={
                            Object {
                              "color": undefined,
                            }
                          }
                          viewBox="0 0 24 24"
                          width="1em"
                          xmlns="http://www.w3.org/2000/svg"
                        >
                          <path
                            d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
                            key="0"
                          />
                        </svg>
                      </IconBase>
                    </MdSearch>
                  </span>
                </Search>
              </span>
            </button>
          </Button>
        </div>
      </form>
    </Formik>
    <CSSTransition
      classNames="address-info"
      in={false}
      timeout={300}
      unmountOnExit={true}
    >
      <Transition
        appear={false}
        enter={true}
        exit={true}
        in={false}
        mountOnEnter={false}
        onEnter={[Function]}
        onEntered={[Function]}
        onEntering={[Function]}
        onExit={[Function]}
        onExited={[Function]}
        onExiting={[Function]}
        timeout={300}
        unmountOnExit={true}
      />
    </CSSTransition>
  </SearchForm>
</AppProvider>
`;
